src/lib.rs

```diff
@@ -435,21 +435,22 @@ where
     ///     println!("done!");
     /// }
     /// ```
-    pub async fn next(&mut self) -> Option<<Self as PollNext>::Item> {
-        return Next(self).await;
+    pub fn next(&mut self) -> Next<'_, Self> {
```

Comment:
I think I have a slight preference for `-> impl Future<Output = Option<<Self as PollNext>::Item>> + FusedFuture + 'a` here, even if it is a bit wordy. In practice, I don't think it matters much since users of the type don't need to type it. It will clutter the docs, I guess.

In general I prefer to reveal as few implementation details as possible since that preserves more semver-compatible options for making changes in the future. 

That said, I would be surprised if this ever changes, and even if it does, once Rust has type-alias-impl-trait, we can just redefine `Next` as `struct Next<'a, T>(impl Future<Output = Option<<Self as PollNext>::Item>> + FusedFuture + 'a)`.

Anyway, I have a slight preference for the `-> impl` version, but I don't think it matters much so I'm fine with either way.

Reply:
Note that another caveat of that approach is that the `FusedFuture` bound would have to be removed when the `futures-rs` feature isn't enabled, and I think the only way of doing that would be to have an entirely different method signature in that case.

Reply:
Oh yeah, good point. I'd say that's a point in favor of returning `Next` then.